### DISCUSSION TITLE
Consider adding `Force` effect in lazy context

### DISCUSSION THREAD

**Discussion by jaschdoc** [flix/flix#5664] [Ideas]
Working on https://github.com/flix/flix/pull/4984 got me thinking we should have a `Force` effect for lazy computations.

Lazy expressions are of course pure but forcing them does have a something close to a side-effect of actually performing a (possibly) larger computation than just simple bookkeeping associated with laziness.
We have a helper function `forceAll: DelayMap[k, v] -> Unit` that forces the backing tree in parallel. This is used when an effectful function is applied to the tree so the evaluation happens faster and we can apply the effectful function sequentially. The `forceAll` function is pure so simply writing `forceAll(m);` doesn't work as the compiler will tell you that it's a useless expression. Then the next solution is to abuse `let _ = forceAll(m);` but this may be unintentionally optimized away.

I think the solution here would be to introduce a `Force` effect that is treated like the `debug` effect where it's pure but not removed by the compiler. `Force` could possibly be allowed in pure contexts (haven't thought this through entirely) since it doesn't affect the end result. Additionally, this would allow a programmer to signal where a lazy computation would actually materialize (I don't know... maybe a library author or contributor?)

Thoughts? Is there maybe a simpler solution to this problem?

**Comment by JonathanStarup** [comment]
In benchmarking there is the same sort of issue. Suddenly you want some arbitrary function to be impure because its effectful on the duration that you are timing.

In delayMap these force functions are really used purely right? in the sense that if you delete them nothing is different except worse performance?

**Comment by JonathanStarup** [comment]
@paulbutcher has talked about a "please optimizer don't touch this" construct and perhaps that is a less complex solution to these problems since they don't actually want anything that the effect system gives.

**Comment by jaschdoc** [comment]
> In benchmarking there is the same sort of issue. Suddenly you want some arbitrary function to be impure because its effectful on the duration that you are timing.


> Paul has talked about a "please optimizer don't touch this" construct and perhaps that is a less complex solution to these problems since they don't actually want anything that the effect system gives.

Come to think of it, this also applies to security-focused programs where you'd want constant time evaluation where it is essential to tell the compiler to not touch that area of the program (however, could the JVM possibly do something to that during runtime?). I'm starting to think this might be a feature we need. What about something simple like `literal exp`?

> In delayMap these force functions are really used purely right? in the sense that if you delete them nothing is different except worse performance?

Correct.

**Comment by paulbutcher** [comment]
> however, could the JVM possibly do something to that during runtime?

As it happens, I used to be Technology Development Manager of Jeode, a non-Oracle JVM which tried to add real-time functionality. So I can say with some authority that yes, the JVM could (in fact is virtually guaranteed to) do something to that at runtime.

There have been some attempts to make a realtime Java:

https://www.oracle.com/technical-resources/articles/javase/jsr-1.html

But they're basically moribund. You can just about create something soft real-time in Java if you try very hard. Hard real-time is a lost cause.

**Comment by JonathanStarup** [comment]
what do you think @magnus-madsen? Are these usecases enough to expose the mask expression?

**Comment by magnus-madsen** [comment]
I am not sure. 

I think the whole point of the effect system is to separate pure and impure code.
To do so in a sound way,
And to allow the compiler to take advantage of that information.

**Comment by JonathanStarup** [comment]
Hmm.. That makes me think: The problem is that the concept of impure is sometimes locally stricter than the normal interpretation. This means that
1. We want locally impure things that
2. stay pure from the outside perspective 

But we can already almost achieve this with regions
```
def delayMapThing(...): ... \ {} = region optimization {
    ....
    forceAll(...) as \ optimization;
    ....
}
```

**Comment by magnus-madsen** [comment]
Interesting thought. Even better the upcast to `optimization` is sound.

**Comment by JonathanStarup** [comment]
> Interesting thought. Even better the upcast to `optimization` is sound.

It is a bit dangerous though since its not an upcast
```
def delayMapThing(): String \ {} = region optimization {
    forceAll() as \ optimization;
    "hey"
}
def forceAll(): Unit \ IO = ??? as \ IO
```

This is required to make it safe
```
def delayMapThing(): String \ {} = region optimization {
    (upcast(forceAll)()): _ \ optimization;
    "hey"
}
def forceAll(): Unit \ IO = ??? as \ IO
```

**Comment by magnus-madsen** [comment]
The more I think about it, the less convinced I am that we should even have `Lazy[t]`.
